src/pss.rs

```diff
@@ -0,0 +1,384 @@
+    //    stop.
+    if db[0] & (0xFF << /*uint*/(8 - (8 * em_len - em_bits))) != 0 {
+        return Err(Error::Verification);
+    }
+
+    // 7. Let dbMask = MGF(H, em_len - h_len - 1)
+    //
+    // 8. Let DB = maskedDB \xor dbMask
+    mgf1_xor(db, hash, &*h);
+
+    // 9.  Set the leftmost 8 * emLen - emBits bits of the leftmost octet in DB
+    //     to zero.
+    db[0] &= 0xFF >> /*uint*/(8 * em_len - em_bits);
+
+    let s_len = match s_len {
+        None => (0..=em_len - (h_len + 2))
+            .rev()
+            .try_fold(None, |state, i| match (state, db[em_len - h_len - i - 2]) {
+                (Some(i), _) => Ok(Some(i)),
+                (_, 1) => Ok(Some(i)),
+                (_, 0) => Ok(None),
+                _ => Err(Error::Verification),
+            })?
+            .ok_or(Error::Verification)?,
+        Some(s_len) => {
+            // 10. If the emLen - hLen - sLen - 2 leftmost octets of DB are not zero
+            //     or if the octet at position emLen - hLen - sLen - 1 (the leftmost
+            //     position is "position 1") does not have hexadecimal value 0x01,
+            //     output "inconsistent" and stop.
+            for e in &db[..em_len - h_len - s_len - 2] {
+                if *e != 0x00 {
+                    return Err(Error::Verification);
+                }
+            }
+            if db[em_len - h_len - s_len - 2] != 0x01 {
+                return Err(Error::Verification);
+            }
+            s_len
+        }
+    };
+
+    // 11. Let salt be the last s_len octets of DB.
+    let salt = &db[db.len() - s_len..];
+
+    // 12. Let
+    //          M' = (0x)00 00 00 00 00 00 00 00 || mHash || salt ;
+    //     M' is an octet string of length 8 + hLen + sLen with eight
+    //     initial zero octets.
+    //
+    // 13. Let H' = Hash(M'), an octet string of length hLen.
+    let prefix = [0u8; 8];
+
+    hash.input(&prefix[..]);
+    hash.input(m_hash);
+    hash.input(salt);
+    let h0 = hash.result_reset();
+
+    // 14. If H = H', output "consistent." Otherwise, output "inconsistent."
+    if Into::<bool>::into(h0.ct_eq(h)) {
+        Ok(())
+    } else {
+        Err(Error::Verification)
+    }
+}
+
+fn inc_counter(counter: &mut [u8]) {
+    if counter[3] == u8::max_value() {
+        counter[3] = 0;
+    } else {
+        counter[3] += 1;
+        return;
+    }
+
+    if counter[2] == u8::max_value() {
+        counter[2] = 0;
+    } else {
+        counter[2] += 1;
+        return;
+    }
+
+    if counter[1] == u8::max_value() {
+        counter[1] = 0;
+    } else {
+        counter[1] += 1;
+        return;
+    }
+
+    if counter[0] == u8::max_value() {
+        counter[0] = 0u8;
+        counter[1] = 0u8;
+        counter[2] = 0u8;
+        counter[3] = 0u8;
+    } else {
+        counter[0] += 1;
+    }
+}
+
+/// Mask generation function
+///
+/// Will reset the Digest before returning.
+fn mgf1_xor(out: &mut [u8], digest: &mut dyn DynDigest, seed: &[u8]) {
```

Comment:
Replace this with the utility function (per earlier comment).

Reply:
fixed